internal/pkg/events/handlers.go

```diff
@@ -0,0 +1,122 @@
+//in future for loading all handler plugins). The plugins will be organize
+//per data source on which's events they could be applied
+type EventHandlerManager struct {
+	Handlers map[saconfig.DataSource][]EventHandler
+}
+
+//NewEventHadlerManager loads all even handler plugins stated in events configuration
+func NewEventHadlerManager(config saconfig.EventConfiguration) (*EventHandlerManager, error) {
+	manager := EventHandlerManager{}
+	manager.Handlers = make(map[saconfig.DataSource][]EventHandler)
+	for _, ds := range []saconfig.DataSource{saconfig.DataSourceCollectd, saconfig.DataSourceCeilometer, saconfig.DataSourceUniversal} {
+		manager.Handlers[ds] = make([]EventHandler, 0)
+	}
+
+	for _, pluginPath := range config.HandlerPlugins {
+		var ds saconfig.DataSource
+		if ok := ds.SetFromString(pluginPath.DataSource); !ok {
+			return &manager, fmt.Errorf("Unknown datasource ''%s' for given event handler", pluginPath.DataSource)
+		}
+		manager.LoadHandlers(ds, pluginPath.Path)
+	}
+
+	//TODO: this just manually register the only handler we have now. Remove when the handler implementation will move out to plugin
+	manager.Handlers[saconfig.DataSourceCollectd] = append(manager.Handlers[saconfig.DataSourceCollectd], ContainerHealthCheckHandler{"collectd_checks"})
+	return &manager, nil
+}
+
+//LoadHandlers will load handler plugins in future
+func (hand *EventHandlerManager) LoadHandlers(dataSource saconfig.DataSource, path string) error {
+
+	return nil
+}
+
+//ContainerHealthCheckHandler serves as handler for events from collectd-sensubility's
+//results of check-container-health.
+type ContainerHealthCheckHandler struct {
+	ElasticIndex string
+}
+
+type containerHealthCheckItem struct {
+	Container string `json:"container"`
+	Service   string `json:"service"`
+	Status    string `json:"status"`
+	Healthy   int    `json:"healthy"`
+}
+
+//Handle saves the event as separate document to ES in case the result output contains more than one item.
+//Returns true if event processing should continue (eg. event should be saved to ES) or false if otherwise.
+func (hand ContainerHealthCheckHandler) Handle(event incoming.EventDataFormat, elasticClient *saelastic.ElasticClient) (bool, error) {
+	if evt, ok := event.(*incoming.CollectdEvent); ok {
+		rawData := evt.GetRawData()
+		if data, ok := rawData.(map[string]interface{}); ok {
+			if rawAnnot, ok := data["annotations"]; ok {
+				if annotations, ok := rawAnnot.(map[string]interface{}); ok {
+					if output, ok := annotations["output"]; ok {
+						var outData []containerHealthCheckItem
+						if err := json.Unmarshal([]byte(output.(string)), &outData); err == nil {
+							// surrogate output key with just one item and save it to ES
+							for _, item := range outData {
+								singleOut, err := json.Marshal(item)
+								if err == nil {
+									annotations["output"] = string(singleOut)
+									_, err = elasticClient.Create(hand.ElasticIndex, EVENTSINDEXTYPE, rawData)
```

Comment:
Gonna have to buy a wider monitor to look at this 😆

Reply:
Unmarshalling the json message to a golang object and then validating that object with https://github.com/go-playground/validator would make this a lot prettier

Reply:
Not sure what do you mean by that. The purpose of the handler is not validate the output data, but to make and save N events to Elasticsearch, where N is count of items in the event output. How am I supposed to do that with the validator?

Reply:
No, I am referring to all of the nested 'if' statements checking that the key and type exists. Instead, if you just unmarshal the original message json into a struct and then use the validator library to validate that those attributes exist, that would already guarantee the existing key and type without having to check everything individually. It would be a lot prettier and more readable.

Reply:
Ah ok I understand now. Well using the arbitrary map of interfaces has one reason here. Having the whole nesting and care just about the output brings compatibility of various sensubility versions. I bet that during time we will change/add/remove data in sensubility messages in near future and unmarshalling to the struct would mean that with every change on sensubility side, we would have to keep in mind that we have to refactor and rebuild the handler.

Using the arbitrary struct of empty interfaces and the not so nice nesting means that the only thing we need to care about regarding this handler is to keep the output value as a member of annotation value and that's it.

Also unmarshalling to defined struct runs validation on the values out of the box AFAIK [1], so even if we would go that way I don't think it's a good idea to bring in another dependency just for additional explicit validation. Or am I wrong here?

[1] https://play.golang.org/p/TA4mPp-WUyp

Reply:
That's a good point, compatability would be more of a problem in the future. Okay, I am convinced.

The type is already determined by golang's type system when json is unmarshalled but nothing will happen if that key does not exist, so you don't have to manually check that with an 'if' statement. The validator library you can use to guarantee a key/value exists by adding the `validate:"required"`. When you run validator.Validate(), it returns an error if, say, the Annotations.Output field does not exist and thus no runtime exception would happen. I used this for the ci-server config: https://github.com/pleimer/ci-server-go/blob/cf6f0f7f665cba122cf6475883a0da4e4584203b/pkg/config/config.go#L18. 

Maybe this is overengineered, but what about a recursive solution? Something like:

```go
type List struct {
	Next *List
	Key string
}

func (hand ContainerHealthCheckHandler) getOutput(path List, data map[string]interface{}) (map[string]interface{}, error) {
	outputs := map[string]interface{}	
	if path.Key == "output" {
		if output, ok := data[path.Key].(map[string]interface{}); ok {
			return output
		}
		return nil, fmt.Errorf("output should be of type []map[string]interface{}")
	}
	
	if _, ok := data[path.Key].(map[string]interface{}); ok {
		return getOutput(path.Next, data[path.Key])
	}
	return nil, fmt.Errorf("input data does not contain path")
}

func (hand ContainerHealthCheckHandler) Handle(event incoming.EventDataFormat, elasticClient *saelastic.ElasticClient) (bool, error) {
	pathList := *List{
		Key: "annotations",
	}
	pathList.Next = *List{
		Key: "output"
	}

	output, err := getOutput(pathList, evt.GetRawData())
	if err != nil {
		return err
	}

	for _, item := range output {
		//set rawData["annotations"]["output"] = item and write to ES
		//there would be no need to check that these values exist and are the right type since getOutput() already did that
	}

}
```

I'm not sure if this is really all that much easier to read, but at least I don't need my ultra-wide monitor to look at it 😀